epic-sample-script.ps1

```diff
@@ -0,0 +1,294 @@
+param
+(   
+    [Parameter(mandatory = $true)]
+    [string] $tenantId,
+    [Parameter(mandatory = $true)]
+    [string] $appId,
+    [Parameter(mandatory = $true)]
+    [string] $appSecret,
+    [Parameter(mandatory = $true)]
+    [string] $jobId,
+    [Parameter(mandatory = $true)]
+    [string] $FilePath,
+    [Parameter(mandatory = $false)]
+    [Int] $RecordsPerCall = 50000
+)
+# Access Token Config
+$oAuthTokenEndpoint = "https://login.windows.net/$tenantId/oauth2/token"
+$resource = 'https://microsoft.onmicrosoft.com/4e476d41-2395-42be-89ff-34cb9186a1ac'
+
+# Csv upload config
+$eventApiURl = "https://webhook.ingestion.office.com"
+$eventApiEndpoint = "api/signals"
+
+$serviceName = "PushConnector"
+
+class FileMetdata {
+    [string]$FileHash
+    [string]$NoOfRowsWritten
+    [string]$Service
+}
+
+function GetAccessToken () {
+    # Token Authorization URI
+    $uri = "$($oAuthTokenEndpoint)?api-version=1.0"
+
+    # Access Token Body
+    $formData = 
+    @{
+        client_id     = $appId;
+        client_secret = $appSecret;
+        grant_type    = 'client_credentials';
+        resource      = $resource;
+        tenant_id     = $tenantId;
+    }
+
+    # Parameters for Access Token call
+    $params = 
+    @{
+        URI         = $uri
+        Method      = 'Post'
+        ContentType = 'application/x-www-form-urlencoded'
+        Body        = $formData
+    }
+
+    $response = Invoke-RestMethod @params -ErrorAction Stop
+    return $response.access_token
+}
+
+function RetryCommand {
+    [CmdletBinding()]
+    Param(
+        [Parameter(Position = 0, Mandatory = $true)]
+        [scriptblock]$ScriptBlock,
+
+        [Parameter(Position = 1, Mandatory = $false)]
+        [int]$Maximum = 15
+    )
+
+    Begin {
+        $cnt = 0
+    }
+
+    Process {
+        do {
+            $cnt++
+            try {
+                $ScriptBlock.Invoke()
+                return
+            }
+            catch {
+                Write-Error $_.Exception.InnerException.Message -ErrorAction Continue
+                Start-Sleep 60
+                if ($cnt -lt $Maximum) {
+                    Write-Output "Retrying"
+                }
+            }
+            
+        } while ($cnt -lt $Maximum)
+
+        throw 'Execution failed.'
+    }
+}
+
+function WriteErrorMessage($errorMessage) {
+    $Exception = [Exception]::new($errorMessage)
+    $ErrorRecord = [System.Management.Automation.ErrorRecord]::new(
+        $Exception,
+        "errorID",
+        [System.Management.Automation.ErrorCategory]::NotSpecified,
+        $TargetObject
+    )
+    $PSCmdlet.WriteError($ErrorRecord)
+}
+
+function UploadData ($access_token, $FileName) {
```

Comment:
Good to annotate top level function with usage information. See https://github.com/microsoft/CSS-Exchange/blob/main/Security/src/Test-ProxyLogon.ps1